### PR TITLE
JS: Add common interface between TypeExpr and JSDocTypeExpr

### PR DESCRIPTION
**Pull request by asger-semmle**
Adds a class `TypeAnnotation` with `TypeExpr` and `JSDocTypeExpr` as subclasses.

It has some methods for common tasks but currently doesn't a mirror hierarchy of the two class hierarchies. I'm not sure how much value we get out of modelling this hierarchy.

For now I'd like to feedback on whether this design is something we can all agree on.

The return type of `getTypeAnnotation()` and friends have changed to `TypeAnnotation`, which is a breaking change since this is not an `ASTNode`. I've copied over some of the predicates from `ASTNode` to reduce the impact. Short of deprecating and renaming all of them, or changing JSDocTypeExpr to be an ASTNode, I think it's unavoidable but low impact.

Still needs an upgrade script, and an evaluation is running in case changing the rootdef of `getEnclosingStmt` and friends have catastrophic consequences.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Variables.qll [hunk 1]
```diff
@@ -501,7 +501,11 @@ class VariableDeclarator extends Expr, @vardeclarator {
   Expr getInit() { result = this.getChildExpr(1) }
 
   /** Gets the TypeScript type annotation for the declared variable or binding pattern. */
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Variables.qll]:
```suggestion
  /** Gets the type annotation for the declared variable or binding pattern. */
```

File: javascript/ql/src/semmle/javascript/TypeAnnotations.qll [hunk 2]
```diff
@@ -0,0 +1,104 @@
+/**
+ * Provides classes for reasoning about type annotations independently of dialect.
+ */
+
+import javascript 
+
+/**
+ * A type annotation, either in the form of a TypeScript type or a JSDoc comment.
+ */
+class TypeAnnotation extends @type_annotation {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/TypeAnnotations.qll]: I suppose we could make this extend `Locatable`?

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/TypeAnnotations.qll]: Done, although the location of the JSDoc annotations seems to be `file://:0:0:0:0`. I'll try and see if there's a simple fix for that.

Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/TypeAnnotations.qll]: > the location of the JSDoc annotations seems to be file://:0:0:0:0

That doesn't sound intentional. If you can fix it that would be great, but of course it's orthogonal to this PR.

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/TypeAnnotations.qll]: I think I fixed it. I'll just do it in the same PR.

**Pull request merged**